02-microarray/pathway-analysis_microarray_04_gsva.Rmd

```diff
@@ -0,0 +1,631 @@
+) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id)) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Now let's join the rest of the expression data
+  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
+```
+
+This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `mapped_df`.
+
+```{r}
+head(mapped_df)
+```
+
+We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
+In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.
+
+Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dup_entrez_ids <- mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+
+dup_entrez_ids
+```
+
+Now let's take a look at the rows associated with the duplicated Entrez IDs.
+
+```{r}
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids)
+```
+
+We can see that the associated values vary for each row.
+
+As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the larger mean expression value per sample.
+
+We are removing values for three genes here, so it is unlikely to have a considerable impact on our results.
+
+**REVIEW**
+
+```{r}
+# First let's create a column with the mean expression value for each row -- we
+# will want to calculate the mean using the numeric expression values only so
+# here we use negative indexing to account for the two gene identifier columns
+# from the previous mapping step that we do not want included in the calculation
+# of the mean
+mapped_df$mean <- apply(mapped_df[, -c(1:2)], 1, mean)
+
+# Let's look at the rows associated with the duplicated Entrez IDs now that we
+# have a mean expression value
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
+  dplyr::select(Ensembl, entrez_id, mean, dplyr::everything())
+```
+
+Now we can filter out the duplicate gene identifiers using the mean expression value.
+
+```{r}
+filtered_mapped_df <- mapped_df %>%
+  # Sort so that the highest absolute mean expression values are at the top
```

Comment:
Why absolute value?

Reply:
Good question, this decision was made based on a [similar decision](https://github.com/AlexsLemonade/refinebio-examples/pull/345#discussion_r517329319) made in the GSEA example where it was suggested that we keep the higher absolute value for the purpose of ranking.

Although GSVA does not require a pre-ranked gene list like GSEA, this decision seemed to still make sense here as the row/observation with the larger absolute value would most likely be more interesting that one with a smaller value (whether it be a positive or negative value).

Reply:
I bet you will respond to https://github.com/AlexsLemonade/refinebio-examples/pull/352#discussion_r524299264 shortly, but some thoughts:

* By using the mean value across samples, you're making a decision based on all the samples but the GSVA scores are on a per-sample basis (part of the gist of the linked comment ☝️ )
* In general, the values that come out of refine.bio for quantile normalized microarray data (what you have here) are going to range from small negative values to higher positive values, where the small negative numbers are genes that are not highly expressed in a particular sample.

Reply:
Gotcha 👍